src/hdmf/common/hierarchicaltable.py

```diff
@@ -0,0 +1,258 @@
+"""
+Module providing additional functionality for dealing with hierarchically nested tables, i.e.,
+tables containing DynamicTableRegion references.
+"""
+import pandas as pd
+import numpy as np
+from hdmf.common.table import DynamicTable, DynamicTableRegion, VectorIndex
+from hdmf.utils import docval, getargs
+
+
+@docval({'name': 'dynamic_table', 'type': DynamicTable,
+         'doc': 'DynamicTable object to be converted to a hierarchical pandas.Dataframe'},
+        returns="Hierarchical pandas.DataFrame with usually a pandas.MultiIndex on both the index and columns.",
+        rtype='pandas.DataFrame',
+        is_method=False)
+def to_hierarchical_dataframe(dynamic_table):
+    """
+    Create a hierarchical pandas.DataFrame that represents all data from a collection of linked DynamicTable.
+
+    **LIMITATIONS:** Currently this function only supports DynamicTables with a single DynamicTableRegion column.
+    If a table has more than one foreign DynamicTableRegion column then the function will expand only the
+    first DynamicTableRegio column found for each table. Any additional DynamicTableRegion columns will remain
+    nested.
+
+    **NOTE:** Some useful functions for further processing of the generated
+    DataFrame include:
+
+    * pandas.DataFrame.reset_index to turn the data from the pandas.MultiIndex into columns
+    * :py:meth:`~hdmf.common.hierarchicaltable.drop_id_columns` to remove all 'id' columns
+    * :py:meth:`~hdmf.common.hierarchicaltable.flatten_column_index` to flatten the column index
+    """
+    # TODO: Need to deal with the case where we have more than one DynamicTableRegion column in a given table
+    # Get the references column
+    foreign_columns = dynamic_table.get_foreign_columns()
+    if len(foreign_columns) == 0:
+        return dynamic_table.to_dataframe()
+    hcol_name = foreign_columns[0]  # We only denormalize the first foreign column for now
+    hcol = dynamic_table[hcol_name]
+    # If our foreign column is a vectorindex
+    hcol_target = hcol.table if isinstance(hcol, DynamicTableRegion) else hcol.target.table
+
+    # Create the data variables we need to collect the data for our output dataframe and associated index
+    index = []
+    data = []
+    columns = None
+    index_names = None
+
+    # If we have indexed columns (other than our hierarchical column) then our index data for our
+    # MultiIndex will contain lists as elements (which are not hashable) and as such create an error.
+    # As such we need to check if we have any affected columns so we can  fix our data
+    # indexed_column_indicies = np.where([isinstance(dynamic_table[colname], VectorIndex)
+    #                                    for colname in dynamic_table.colnames if colname != hcol_name])[0]
+    # indexed_column_indicies += 1  # Need to increment by 1 since we add the row id in our iteration below
```

Comment:
Does this still need to be here? If so, please add a comment as to why it is commented out and how that changes/limits the behavior of the function if at all.

Reply:
I think this can go. As I was going through testing I realized that this was actually not used. And I honestly couldn't come up with a case where this would be still needed.

Reply:
I remove the commented code.